src/Microsoft.SqlTools.ServiceLayer/EditData/SmoEditMetadataFactory.cs

```diff
@@ -0,0 +1,69 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Data.Common;
+using System.Data.SqlClient;
+using Microsoft.SqlServer.Management.Common;
+using Microsoft.SqlServer.Management.Smo;
+using Microsoft.SqlTools.ServiceLayer.Connection.ReliableConnection;
+using Microsoft.SqlTools.ServiceLayer.EditData.UpdateManagement;
+using Microsoft.SqlTools.ServiceLayer.QueryExecution.Contracts;
+
+namespace Microsoft.SqlTools.ServiceLayer.EditData
+{
+    /// <summary>
+    /// Factory that generates metadata using a combination of SMO and SqlClient metadata
+    /// </summary>
+    public class SmoEditMetadataFactory : IEditMetadataFactory
+    {
+        /// <summary>
+        /// Generates a edit-ready metadata object using SMO
+        /// </summary>
+        /// <param name="connection">Connection to use for getting metadata</param>
+        /// <param name="columns">List of columns from a query against the object</param>
+        /// <param name="objectName">Name of the object to return metadata for</param>
+        /// <param name="objectType">Type of the object to return metadata for</param>
+        /// <returns>Metadata about the object requested</returns>
+        public IEditTableMetadata GetObjectMetadata(DbConnection connection, DbColumnWrapper[] columns, string objectName, string objectType)
+        {
+            // Get a connection to the database for SMO purposes
+            SqlConnection sqlConn = connection as SqlConnection;
+            if (sqlConn == null)
+            {
+                // It's not actually a SqlConnection, so let's try a reliable SQL connection
+                ReliableSqlConnection reliableConn = connection as ReliableSqlConnection;
+                if (reliableConn == null)
+                {
+                    // If we don't have connection we can use with SMO, just give up on using SMO
+                    return null;
+                }
+
+                // We have a reliable connection, use the underlying connection
+                sqlConn = reliableConn.GetUnderlyingConnection();
+            }
+
+            Server server = new Server(new ServerConnection(sqlConn));
+            TableViewTableTypeBase result;
+            switch (objectType.ToLowerInvariant())
+            {
+                case "table":
+                    result = server.Databases[sqlConn.Database].Tables[objectName];
+                    break;
+                case "view":
+                    result = server.Databases[sqlConn.Database].Views[objectName];
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(objectType), SR.EditDataUnsupportedObjectType(objectType));
```

Comment:
I think I can count on one hand how many types of exceptions are being thrown. Either way, I'm not sure that using less exceptions is what we'd want to do anyhow. You'd have to give me a better example of what you mean.